cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextProvider.java

```diff
@@ -98,46 +99,43 @@ private synchronized AbstractMapReduceContextBuilder getBuilder(CConfiguration c
     return contextBuilder;
   }
 
-  private boolean isLocal() {
-    String mrFramework = taskContext.getConfiguration().get(MRConfig.FRAMEWORK_NAME, MRConfig.LOCAL_FRAMEWORK_NAME);
+  private boolean isLocal(Configuration hConf) {
+    String mrFramework = hConf.get(MRConfig.FRAMEWORK_NAME, MRConfig.LOCAL_FRAMEWORK_NAME);
     return "local".equals(mrFramework);
   }
 
-  private Program createProgram() {
-    // If Program is already created, return it.
-    // It is needed so that in distributed mode, there is only one ProgramClassLoader created, even if
-    // there are multiple instances of this class being created (one from creating RecordReader/Writer, one from
-    // MapperWrapper/ReducerWrapper).
-
-    // For local mode, it doesn't matter than much, but no harm in returning the same program.
-    Program program = PROGRAM_THREAD_LOCAL.get();
-    if (program != null) {
-      return program;
+  private Program createProgram(MapReduceContextConfig contextConfig) {
+    LocationFactory locationFactory = new LocalLocationFactory();
```

Comment:
Could we just hang this final member variable? I assume the value will be the same and no state would be change during the lifetime of the context provider?